### ISSUE TITLE
[8.0] Allow using http_auth with RequestsHttpNode for custom authentication

### ISSUE THREAD

**Issue by sethmlarson** [elastic/elasticsearch-py#1778]
Using `http_auth` is deprecated for passing `Tuple[str, str]` and `str` but is useful for custom HTTP authentication with the Requests library. For example Kerberos and GSS are possible with this method. To allow for these to continue working we'll continue setting `requests.Session.auth = http_auth` when used this way.

Reference: https://github.com/elastic/elasticsearch-py/pull/1776#issuecomment-953299419

**Comment by sethmlarson** [comment]
Work is done on the transport side: https://github.com/elastic/elastic-transport-python/pull/48

**Comment by naisanzaa** [comment]
Error importing `RequestsHttpConnection`:

```
from elasticsearch import RequestsHttpConnection

Traceback (most recent call last):
  File "/Users/eric/venv/garbage/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3251, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-8-6f544e924cdd>", line 1, in <module>
    from elasticsearch import RequestsHttpConnection
ImportError: cannot import name 'RequestsHttpConnection' from 'elasticsearch' (/Users/eric/venv/garbage/lib/python3.9/site-packages/elasticsearch/__init__.py)

```

```
elastic-transport==8.0.1
elasticsearch==8.0.0

**Comment by sethmlarson** [comment]
@naisanzaa Transport-layer classes aren't supplied via the `elasticsearch` package any more, the `elastic_transport` package is used for multiple packages including `elasticsearch`. You can find `RequestsHttpNode` in that package.

**Comment by sethmlarson** [comment]
Since there's a lot of interest in this feature, this will be released in 8.2.0 but support is currently merged into `main`.

**Comment by james-henry-git** [comment]
> So what is the solution to this problem ? I'm still having the same issue, with ES 8.1.0 This [stackoverflow post](https://stackoverflow.com/questions/71182063/cannot-import-name-requestshttpconnection-from-elasticsearch/71334968#71334968) has an answer that says to downgrade from python package elasticsearch==8.0.0 to elasticsearch==7.13.4

FYI, for those struggling with using the requests module in Elasticsearch 8.x is hardly documented well, we shouldn't need to thunk down versions, we just need to tell it to use RequestsHttpNode and set the appropriate variables.  

```
import os
os.environ["https_proxy"] = "http://proxy_name:8080"
from elastic_transport import RequestsHttpNode
from elasticsearch import Elasticsearch
client = Elasticsearch(
    cloud_id = authkeys['cloud_id'], 
    api_key = authkeys['api_key'],
    node_class=RequestsHttpNode,
    )
```

**Comment by nartix** [comment]
Here is a simple fix if you want to pass custom header fields in an Elasticsearch connection in the latest version.

```
from elasticsearch_dsl import connections
from elasticsearch.transport import Transport

class HeaderInjectingTransport(Transport):
    def perform_request(self, method, url, headers=None, body=None,  *args, **kwargs):
        if headers is None:
            headers = {}

        headers['custom-header-field'] = "custom data"

        return super(HeaderInjectingTransport, self).perform_request(
            method, url, headers=headers, body=body, *args, **kwargs)

        connections.create_connection(
            hosts=["server_address],
            transport_class=HeaderInjectingTransport
        )
```

**Comment by pquentin** [comment]
Hey @Adam-Sparrow, there's a lot to unpack here, so please bear with me.

> They had us install version 7.14, but the issue persisted in my instance. That is likely my fault, but still it seems I am being taught something out of date for at least 2.5 years.

Correct, Elasticsearch 7.14 was released in August 2021 and is no longer maintained or supported. Elasticsearch 7.17 will stop being maintained soon, as we're working on 9.0.0. https://www.elastic.co/support/eol#prior-versions contains more details. I would suggest using the latest 8.x now and then switching to 9.x when it's released. Especially for a course where by definition there's no pre-existing running cluster.

That said, I acknowledge that elasticsearch-py 8.0 introduced a few breaking changes as well as warnings, and elasticsearch-py 9.0 will turn those warnings into exceptions.

> I also don't have cloud_id and an api_key to my knowledge unless that is my user and password secrets.

`cloud_id` is specific to Elastic Cloud. It's not a secret, just URLs for Elasticsearch and Kibana encoded as base64. Sticking to a normal URL is perfectly fine. Regarding API keys, they're the preferred way for authentication, as they're more flexible, but for a course using [Basic HTTP Authentication](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication) as you snippet is doing is fine.

> When I use the os.environ["https_proxy"] line, do I need to supply a proxy_name?

No, what's `proxy_name`? While the client does not suport proxies directly, there are workarounds depending on the version:

 * 7.x: https://github.com/elastic/elasticsearch-py/issues/275#issuecomment-218403073
 * 8.x and 9.x: https://github.com/elastic/elastic-transport-python/issues/53#issuecomment-1447903214

> Unfortunately the course materials provide us the import code which is failing in this thread.

The code works if you use Elasticsearch 7.17 elasticsearch-py 7.14, but that depends on what you have in secrets. To run Elasticsearch locally, I recommend the new [start-local](https://www.elastic.co/guide/en/elasticsearch/reference/current/run-elasticsearch-locally.html) tool. To use Elasticsearch 7.17:

```
$ curl -fsSL https://elastic.co/start-local | sh -s -- -v 7.17.28

  ______ _           _   _      
 |  ____| |         | | (_)     
 | |__  | | __ _ ___| |_ _  ___ 
 |  __| | |/ _` / __| __| |/ __|
 | |____| | (_| \__ \ |_| | (__ 
 |______|_|\__,_|___/\__|_|\___|
-------------------------------------------------
🚀 Run Elasticsearch and Kibana for local testing
-------------------------------------------------

ℹ️  Do not use this script in a production environment

⌛️ Setting up Elasticsearch and Kibana v7.17.28...

- Generated random passwords
- Created the elastic-start-local folder containing the files:
  - .env, with settings
  - docker-compose.yml, for Docker services
  - start/stop/uninstall commands
- Running docker compose up --wait

[...]


🎉 Congrats, Elasticsearch and Kibana are installed and running in Docker!

🌐 Open your browser at http://localhost:5601

   Username: elastic
   Password: LYnv0wAe

🔌 Elasticsearch API endpoint: http://localhost:9200
🔑 Use basic auth or create an API key
https://www.elastic.co/guide/en/kibana/current/api-keys.html


Learn more at https://github.com/elastic/start-local
```

Using the following script adapted from your own:

```python
from elasticsearch import Elasticsearch, __version__
from elasticsearch import RequestsHttpConnection


print(__version__)
# Load the secrets
secrets = {
    "host": "localhost",
    "scheme": "http",
    "port": 9200,
    "user": "elastic",
    "pass": "LYnv0wAe",
    "prefix": "/",
}

es = Elasticsearch(
    [secrets["host"]],
    http_auth=(secrets["user"], secrets["pass"]),
    url_prefix=secrets["prefix"],
    scheme=secrets["scheme"],
    port=secrets["port"],
    connection_class=RequestsHttpConnection,
)
print(es.info()["tagline"])
```

(You should 100% keep using `hidden`, but I don't know what it is, so I'm just hardcoding the secrets here.)

Now, in another directory, install elasticsearch-py 7.14 (or 7.17) and requests:

```bash
$ python3 -m venv .venv
$ source .venv/bin/activate
$ pip install elasticsearch==7.14 requests
$ python issues/1778_7.x.py 
(7, 14, 0)
You Know, for Search
```

Now, to use elasticsearch 8.x (8.17.2 at the time of writing) or elasticsearch-py 9.x (not released at the time of writing), we can do the same thing. First, I ran `bash uninstall.sh` in the `elastic-start-local` directory. Then I installed 8.17.2:

```
$ curl -fsSL https://elastic.co/start-local | sh
[...]

🎉 Congrats, Elasticsearch and Kibana are installed and running in Docker!

🌐 Open your browser at http://localhost:5601

   Username: elastic
   Password: BvE2af2A

🔌 Elasticsearch API endpoint: http://localhost:9200
🔑 API key: TTBzaFc1VUJwdld4QzJuUHBUUXU6eHUwS1VIR3lRSC1EWmdyZDlQYUN4UQ==
```

In the virtual environment, I updated to elasticsearch-py 8.x and ran the updated script:

```python
from elasticsearch import Elasticsearch, __version__


print(__version__)
# Load the secrets
secrets = {
    "user": "elastic",
    "pass": "BvE2af2A",
}

es = Elasticsearch(
    "http://localhost:9200",
    basic_auth=(secrets["user"], secrets["pass"]),
    node_class="requests",
)
print(es.info().body["tagline"])
```

```
$ pip install --upgrade elasticsearch
[...]
Successfully installed elastic-transport-8.17.0 elasticsearch-8.17.1
$ python issues/1778_8.x.py
(8, 17, 1)
You Know, for Search
```

Hope that helps!

**Comment by Adam-Sparrow** [comment]
Huge thanks @pquentin I can follow your code and it has lead to a successful result.

In regards to proxy_name, I had no idea 😅 I misread the previous function to believe it was an argument of the elasticsearch client, that was my bad!

I am missing some of the prerequisities for start local, the course I am studying is actually an SQL course and I believe the elasticsearch module is supposed to help us better understand indexing, but we haven't used docker for any of the rest of the course and I don't have access to a Linux VM or have a wsl install running currently. So I decided to avoid using a local elasticsearch instance and rolled back as you suggested. 

We should have access to an elasticsearch instance hosted by our course provider, and eventually I did get access despite further issues related to incorrect credentials passed from the course provider (to give a fake example representing the differences: prefix = "elastic" where it should have been "version/auth/elasticsearch")

Your suggested code was critical to finding these issues, as previously I would just get attribute or API errors, so thank you!

I did end up using !pip install elasticsearch==7.17 over 7.14, 7.14 threw Attribute errors on ElasticSearch import for numpy mismatches. 

I have successfully established a connection to the course elasticsearch instance! I hope that my course providers will also read your suggestions and update their hosted instance and related materials. I will just try to survive the rest of the course.
